### PR TITLE
Implementing solid

### PR DESCRIPTION
**Pull request by CobbyElsonfx**
### Description

This PR addresses the need to clean up and refactor our codebase. As our project has evolved, it has become substantial, with the `main.rb` file particularly growing in size. To ensure code maintainability and alignment with best practices, we're making several improvements.

### Changes Made

1. **Adhering to YAGNI, KISS, and DRY Principles:** We've meticulously reviewed the codebase to ensure that we adhere to the "You Ain't Gonna Need It" (YAGNI), "Keep It Simple, Stupid" (KISS), and "Don't Repeat Yourself" (DRY) principles. This means removing unnecessary code, simplifying complex sections, and eliminating redundancy for more efficient and clean code.

2. **SOLID Principles:** Following SOLID principles has been a central focus. In particular, we've emphasized the "Single-responsibility Principle" to make sure that each class or module has a clear and singular purpose.

3. **Main.rb Role Clarification:** We've refactored the `main.rb` file to focus solely on initiating the code and asking the user for initial options or to quit. By separating concerns, we make the code easier to understand and maintain.

4. **Extracting Logic into Separate Classes:** To achieve better organization and maintainability, we've extracted logical pieces into separate classes or modules. This separation of concerns enhances code readability and future development.

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -1,43 +1,7 @@
+# main.rb
 require_relative 'app'
+require_relative 'user_interface'
 
-class Main
```
Comment (Efiamotu-1) [inline main.rb:3]: - [ ] Following the project requirement you are to make sure that the role of main.rb is only to start their code and ask the user for the initial choice of options (or to quit). Please kindly follow this requrement :+1: :sparkles:

File: README.md
Comment (Efiamotu-1) [inline README.md:1]: - [ ] In your `app.rb` file under the method `create_person` you are not following the single Responsibility Rule.
- [ ] To follow the `Single Responsibility Rule` for best practices all requests for input and input validation should've been placed in the main.rb file and/or classes/methods whose responsibilities are to request or validate input. see the example code snippet below
- code snippets
```ruby
    print "Do you want to create a student (1) or a teacher (2)? [Input the number]: "
    person_type = gets.chomp

    print 'Title?'
    title = gets.chomp

    print 'Author?'
    author = gets.chomp
```
- [ ] Please kindly update your code base in your `app.rb` file to follow this rule by implementing the above requirement everywhere in your app where you are requesting for `user inputs` :+1: :sparkles:

Comment (Efiamotu-1) [inline README.md:1]: - [ ] The next requirement is also in your `app.rb` file under the `create person` method, 
- [ ] this below line of code in your create_student method should've been placed in the `create_person()` method in your `app.rb` file
```ruby
   student = Student.new(classroom, age, name, parent_permission: permission)
    @people.push(student)
```
- [ ] this below line of code in your create_teacher method should've been placed in the `create_person()` method in your `app.rb` file.
```ruby
teacher = Teacher.new(specialization, age, name: name, parent_permission: true)
    @people.push(teacher)
```

**Review by Efiamotu-1 — CHANGES_REQUESTED** [review]
Hi @CobbyElsonfx & @Rafiduzzaman & @78fatemanazari  

 

Good job so far! 👏. Your Code is Awesome! 💪

There are some **few** issues 🐛 that you still need to work 🔧 on though to go to the next project ⏭️ but you are almost there! 🏃
## Don't Worry 🙏

It's ok to have bugs in your code 🐛. Please, see this code review as a **chance** for you to **learn** practical coding skills and learn **new tips** that can improve your code quality 💯 and help you **get a better ** 💼.

_I am going to guide you on how to fix them down below. 🤓_

## Highlights :dart: 
- ✔️ No linter Errors
- ✔️ Descriptive README
- ✔️ Descriptive PR title and description
- The code is top-notch! 💯 and has a huge potential.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


## Important Notes 📝

- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

> **Note:** Please, don't Merge This Pull Request yet. Only Merge when it is Approved ✅.


## Thank you for your time 🙏 and call me if you need help or clarification ☎️

- This is My **GitHub**: [@Efiamotu-1](https://github.com/Efiamotu-1/)
- This is My **Twitter**: [@EFYAMOTU](https://twitter.com/EFYAMOTU)
- This is My **linkedIn**.[Habeeb Efiamotu Musa](https://www.linkedin.com/in/musa-habeeb/)
- This is My **Slack**: [Habeeb Efiamotu Musa](https://slack.com)

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: README.md
Comment (78fatemanazari) [reply README.md:1]: Hi dear @Efiamotu-1 
Can you give me more explanations please?

**Review by skyv26 — APPROVED** [review]
## Hi @CobbyElsonfx and @78fatemanazari, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:  
Congratulations! 🎉 

## STATUS: APPROVED 🟢 



## Highlights
- Good **PR** title ✅ 
- Descriptive *PR* description ✅ 
- Implemented the features asked by previous code reviewer. ✅ 



## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

## ⚠️  WARNING ⚠️ 
**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

## Have any doubt ❓ 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @skyv26 in your question so I can receive the notification. You can also follow me on the below platforms

<table>
<tr>
<td>

<a href="https://github.com/skyv26">
   @skyv26
</a>

</td>
<td>

<a href="https://www.linkedin.com/in/skyv2022/">
   linkedin-skyv2022
</a>

</td>
<td>

<a href="https://twitter.com/vrma_aakash">
   @vrma_aakash
</a>

</td>
<td>

<a href="https://www.instagram.com/skybrel/">
   @skybrel
</a>

</td>
<td>

<a href="https://www.frontendmentor.io/profile/skyv26">
   @skybrel
</a>

</td>
</tr>
</table>

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**